File: go/modcdp/injector/BorrowExtensionInjector.go
// MODCDP_TRANSLATE: KEEP THIS FILE TRANSLATED ACROSS TYPESCRIPT, PYTHON, AND GO.
// Keep all shapes, signatures, behavior, and tests 1:1 in sync with:
// - ./js/src/injector/BorrowExtensionInjector.ts
// - ./python/modcdp/injector/BorrowExtensionInjector.py
package injector

import (
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"time"
)

type BorrowExtensionInjector struct {
	ExtensionInjector
	UnpackedExtensionPath           string
	CleanupPath                     string
	BootstrapModCDPServerExpression string
}

const borrowBootstrapStatusExpression = `(() => ({
  ok: Boolean(globalThis.ModCDP?.handleCommand && globalThis.ModCDP?.addCustomEvent),
  extension_id: globalThis.chrome?.runtime?.id ?? null,
  has_tabs: Boolean(globalThis.chrome?.tabs?.query),
  has_debugger: Boolean(globalThis.chrome?.debugger?.sendCommand && globalThis.chrome?.debugger?.getTargets),
}))()`

type borrowedExtensionCandidate struct {
	result      *ExtensionInjectionResult
	hasTabs     bool
	hasDebugger bool
}

func NewBorrowExtensionInjector(options InjectorOptions) BorrowExtensionInjector {
	return BorrowExtensionInjector{ExtensionInjector: NewExtensionInjector(options)}
}

func (i *BorrowExtensionInjector) Prepare() error {
	if i.BootstrapModCDPServerExpression != "" {
		return nil
	}
	unpackedPath, cleanupPath, err := prepareUnpackedExtension(i.Config.InjectorBorrowExtensionPath)
	if err != nil {
		return err
	}
	i.UnpackedExtensionPath = unpackedPath
	i.CleanupPath = cleanupPath
	body, err := os.ReadFile(filepath.Join(i.UnpackedExtensionPath, "modcdp", "service_worker.js"))
	if err != nil {
		_ = os.RemoveAll(i.CleanupPath)
		i.CleanupPath = ""
		i.UnpackedExtensionPath = ""
		return err
	}
	source := string(body)
	i.BootstrapModCDPServerExpression = fmt.Sprintf(`async function() {
if (!globalThis.ModCDP) {
%s
}
const ModCDP = globalThis.ModCDP;
return {
  ok: Boolean(ModCDP?.handleCommand && ModCDP?.addCustomEvent),
  extension_id: globalThis.chrome?.runtime?.id ?? null,
  has_tabs: Boolean(globalThis.chrome?.tabs?.query),
  has_debugger: Boolean(globalThis.chrome?.debugger?.sendCommand && globalThis.chrome?.debugger?.getTargets),
};
}`, source)
	return nil
}

func (i *BorrowExtensionInjector) Close() error {
	if i.CleanupPath != "" {
		_ = os.RemoveAll(i.CleanupPath)
		i.CleanupPath = ""
	}
	return nil
}

func (i *BorrowExtensionInjector) Inject() (*ExtensionInjectionResult, error) {
	deadline := time.Now().Add(time.Duration(i.Config.InjectorServiceWorkerReadyTimeoutMS) * time.Millisecond)
	for {
		borrowed, err := i.borrowVisibleServiceWorkers()
		if err != nil || borrowed != nil {
			return borrowed, err
		}
		if time.Now().After(deadline) {
			return nil, nil
		}
		time.Sleep(time.Duration(i.Config.InjectorServiceWorkerPollIntervalMS) * time.Millisecond)
	}
}

func (i *BorrowExtensionInjector) borrowVisibleServiceWorkers() (*ExtensionInjectionResult, error) {
	targets, err := i.targetInfos()
	if err != nil {
		return nil, err
	}
	hasConfiguredMatcher := i.Config.InjectorServiceWorkerExtensionID != "" || len(i.Config.InjectorServiceWorkerURLIncludes) > 0 || len(i.Config.InjectorServiceWorkerURLSuffixes) > 0
	candidates := []map[string]any{}
	for _, target := range targets {
		targetType, _ := target["type"].(string)
		targetURL, _ := target["url"].(string)
		if targetType != "service_worker" || !strings.HasPrefix(targetURL, "chrome-extension://") {
			continue
		}
		if hasConfiguredMatcher && !i.serviceWorkerTargetMatches(target) {
			continue
		}
		candidates = append(candidates, target)
	}
	var borrowed []borrowedExtensionCandidate
	for _, target := range candidates {
		bootstrapped, err := i.bootstrapTarget(target)
		if err == nil && bootstrapped != nil {
			borrowed = append(borrowed, *bootstrapped)
		}
	}
	sort.SliceStable(borrowed, func(left, right int) bool {
		if borrowed[left].hasDebugger != borrowed[right].hasDebugger {
			return borrowed[left].hasDebugger
		}
		return borrowed[left].hasTabs && !borrowed[right].hasTabs
	})
	if len(borrowed) == 0 {
		return nil, nil
	}
	return borrowed[0].result, nil
}

func (i *BorrowExtensionInjector) bootstrapTarget(target map[string]any) (*borrowedExtensionCandidate, error) {
	targetID, _ := target["targetId"].(string)
	targetURL, _ := target["url"].(string)
	attached, err := i.sendWithTimeout("Target.attachToTarget", map[string]any{"targetId": targetID, "flatten": true}, "", i.Config.InjectorServiceWorkerProbeTimeoutMS)
	if err != nil {
		return nil, err
	}
	sessionID, _ := attached["sessionId"].(string)
	if sessionID == "" {
		return nil, fmt.Errorf("Target.attachToTarget returned no sessionId for targetId=%s", targetID)
	}
	detach := func() {
		_, _ = i.sendWithTimeout("Target.detachFromTarget", map[string]any{"sessionId": sessionID}, "", i.Config.InjectorCDPSendTimeoutMS)
	}
	_, _ = i.sendWithTimeout("Runtime.enable", map[string]any{}, sessionID, i.Config.InjectorCDPSendTimeoutMS)
	status, err := i.sendWithTimeout("Runtime.evaluate", map[string]any{
		"expression":    borrowBootstrapStatusExpression,
		"returnByValue": true,
	}, sessionID, i.Config.InjectorCDPSendTimeoutMS)
	if err != nil {
		detach()
		return nil, err
	}
	result, _ := status["result"].(map[string]any)
	value, _ := result["value"].(map[string]any)
	if hasTabs, _ := value["has_tabs"].(bool); !hasTabs {
		detach()
		return nil, nil
	}
	if hasDebugger, _ := value["has_debugger"].(bool); !hasDebugger {
		detach()
		return nil, nil
	}
	if ok, _ := value["ok"].(bool); !ok {
		if i.BootstrapModCDPServerExpression == "" {
			detach()
			return nil, fmt.Errorf("BorrowExtensionInjector requires Prepare before Inject")
		}
		probe, err := i.sendWithTimeout("Runtime.evaluate", map[string]any{
			"expression":    fmt.Sprintf("(%s)()", i.BootstrapModCDPServerExpression),
			"awaitPromise":  true,
			"returnByValue": true,
		}, sessionID, i.Config.InjectorCDPSendTimeoutMS)
		if err != nil {
			detach()
			return nil, err
		}
		result, _ = probe["result"].(map[string]any)
		value, _ = result["value"].(map[string]any)
		if hasTabs, _ := value["has_tabs"].(bool); !hasTabs {
			detach()
			return nil, nil
		}
		if hasDebugger, _ := value["has_debugger"].(bool); !hasDebugger {
			detach()
			return nil, nil
		}
	}
	ready, _ := value["ok"].(bool)
	if ready && i.readyExpression() != modcdpReadyExpression {
		readyProbe, err := i.sendWithTimeout("Runtime.evaluate", map[string]any{
			"expression":    i.readyExpression(),
			"returnByValue": true,
		}, sessionID, i.Config.InjectorCDPSendTimeoutMS)
		if err != nil {
			detach()
			return nil, err
		}
		readyResult, _ := readyProbe["result"].(map[string]any)
		ready, _ = readyResult["value"].(bool)
	}
	if !ready {
		detach()
		return nil, nil
	}
	extensionID, _ := value["extension_id"].(string)
	if extensionID == "" {
		if match := extIDFromURL.FindStringSubmatch(targetURL); len(match) > 1 {
			extensionID = match[1]
		}
	}
	hasTabs, _ := value["has_tabs"].(bool)
	hasDebugger, _ := value["has_debugger"].(bool)
	return &borrowedExtensionCandidate{
		result: &ExtensionInjectionResult{
			Source:      "borrow",
			ExtensionID: extensionID,
			TargetID:    targetID,
			URL:         targetURL,
			SessionID:   sessionID,
		},
		hasTabs:     hasTabs,
		hasDebugger: hasDebugger,
	}, nil
}

File: python/modcdp/injector/BorrowExtensionInjector.py
from __future__ import annotations

import tempfile
import time
from pathlib import Path
from typing import Any, Mapping, cast

from ..injector.ExtensionInjector import (
    DEFAULT_SERVICE_WORKER_POLL_INTERVAL_MS,
    DEFAULT_SERVICE_WORKER_PROBE_TIMEOUT_MS,
    DEFAULT_SERVICE_WORKER_READY_TIMEOUT_MS,
    EXT_ID_FROM_URL_RE,
    MODCDP_READY_EXPRESSION,
    ExtensionInjectionResult,
    ExtensionInjector,
    InjectorOptions,
    defaultModCDPExtensionPath,
    prepareUnpackedExtension,
)

BORROW_BOOTSTRAP_STATUS_EXPRESSION = """
(() => ({
  ok: Boolean(globalThis.ModCDP?.handleCommand && globalThis.ModCDP?.addCustomEvent),
  extension_id: globalThis.chrome?.runtime?.id ?? null,
  has_tabs: Boolean(globalThis.chrome?.tabs?.query),
  has_debugger: Boolean(globalThis.chrome?.debugger?.sendCommand && globalThis.chrome?.debugger?.getTargets),
}))()
"""


class BorrowExtensionInjector(ExtensionInjector):
    def __init__(self, options: InjectorOptions | None = None) -> None:
        super().__init__(options)
        self.unpacked_extension_path: str | None = None
        self.cleanup: tempfile.TemporaryDirectory[str] | None = None
        self.bootstrap_modcdp_server_expression: str | None = None

    def prepare(self) -> None:
        if self.bootstrap_modcdp_server_expression is not None:
            super().prepare()
            return
        extension_path = self.options.get("injector_borrow_extension_path") or defaultModCDPExtensionPath()
        if not extension_path:
            raise FileNotFoundError("Unable to locate bundled ModCDP extension for borrow injector.")
        self.unpacked_extension_path, self.cleanup = prepareUnpackedExtension(extension_path)
        try:
            source = (Path(self.unpacked_extension_path) / "modcdp" / "service_worker.js").read_text()
        except BaseException:
            self.cleanup.cleanup()
            self.cleanup = None
            self.unpacked_extension_path = None
            raise
        self.bootstrap_modcdp_server_expression = (
            "async function() {\n"
            "if (!globalThis.ModCDP) {\n"
            f"{source}\n"
            "}\n"
            "const ModCDP = globalThis.ModCDP;\n"
            "return {\n"
            "  ok: Boolean(ModCDP?.handleCommand && ModCDP?.addCustomEvent),\n"
            "  extension_id: globalThis.chrome?.runtime?.id ?? null,\n"
            "  has_tabs: Boolean(globalThis.chrome?.tabs?.query),\n"
            "  has_debugger: Boolean(globalThis.chrome?.debugger?.sendCommand && globalThis.chrome?.debugger?.getTargets),\n"
            "};\n"
            "}"
        )
        super().prepare()

    def close(self) -> None:
        super().close()
        if self.cleanup is not None:
            self.cleanup.cleanup()
            self.cleanup = None

    def inject(self) -> ExtensionInjectionResult | None:
        deadline = time.monotonic() + (self.options.get("injector_service_worker_ready_timeout_ms") or DEFAULT_SERVICE_WORKER_READY_TIMEOUT_MS) / 1000
        while True:
            borrowed = self._borrowVisibleServiceWorkers()
            if borrowed:
                return borrowed
            if time.monotonic() >= deadline:
                return None
            time.sleep((self.options.get("injector_service_worker_poll_interval_ms") or DEFAULT_SERVICE_WORKER_POLL_INTERVAL_MS) / 1000)

    def _borrowVisibleServiceWorkers(self) -> ExtensionInjectionResult | None:
        borrowed: list[tuple[ExtensionInjectionResult, bool, bool]] = []
        visible_service_workers = [
            target
            for target in self._targetInfos()
            if target.get("type") == "service_worker" and isinstance(target.get("url"), str) and target["url"].startswith("chrome-extension://")
        ]
        has_configured_matcher = bool(
            self.options.get("injector_service_worker_extension_id")
            or self.options.get("injector_service_worker_url_includes")
            or self.options.get("injector_service_worker_url_suffixes")
        )
        candidates = [target for target in visible_service_workers if self._serviceWorkerTargetMatches(target)] if has_configured_matcher else visible_service_workers
        for target in candidates:
            try:
                bootstrapped = self._bootstrapTarget(target)
            except Exception:
                bootstrapped = None
            if bootstrapped:
                borrowed.append(bootstrapped)
        borrowed.sort(key=lambda item: (item[2], item[1]), reverse=True)
        return borrowed[0][0] if borrowed else None

    def _bootstrapTarget(self, target) -> tuple[ExtensionInjectionResult, bool, bool] | None:
        attached = self._sendWithTimeout(
            "Target.attachToTarget",
            {"targetId": target["targetId"], "flatten": True},
            None,
            self.options.get("injector_service_worker_probe_timeout_ms") or DEFAULT_SERVICE_WORKER_PROBE_TIMEOUT_MS,
        )
        session_id = attached.get("sessionId")
        if not isinstance(session_id, str) or not session_id:
            raise RuntimeError(f"Target.attachToTarget returned no sessionId for targetId={target['targetId']}")
        try:
            try:
                self._sendWithTimeout("Runtime.enable", {}, session_id)
            except Exception:
                pass
            status = self._sendWithTimeout(
                "Runtime.evaluate",
                {
                    "expression": BORROW_BOOTSTRAP_STATUS_EXPRESSION,
                    "returnByValue": True,
                },
                session_id,
            )
            result = cast(Mapping[str, Any], status.get("result")) if isinstance(status.get("result"), Mapping) else {}
            raw_value = result.get("value")
            value = cast(Mapping[str, Any], raw_value) if isinstance(raw_value, Mapping) else {}
            if not bool(value.get("has_tabs")) or not bool(value.get("has_debugger")):
                self._sendWithTimeout("Target.detachFromTarget", {"sessionId": session_id})
                return None
            if not bool(value.get("ok")):
                if self.bootstrap_modcdp_server_expression is None:
                    raise RuntimeError("BorrowExtensionInjector requires prepare before inject.")
                bootstrap = self._sendWithTimeout(
                    "Runtime.evaluate",
                    {
                        "expression": f"({self.bootstrap_modcdp_server_expression})()",
                        "awaitPromise": True,
                        "returnByValue": True,
                    },
                    session_id,
                )
                result = cast(Mapping[str, Any], bootstrap.get("result")) if isinstance(bootstrap.get("result"), Mapping) else {}
                raw_value = result.get("value")
                value = cast(Mapping[str, Any], raw_value) if isinstance(raw_value, Mapping) else {}
            if not bool(value.get("has_tabs")) or not bool(value.get("has_debugger")):
                self._sendWithTimeout("Target.detachFromTarget", {"sessionId": session_id})
                return None
            ready = bool(value.get("ok"))
            if ready and self._readyExpression() != MODCDP_READY_EXPRESSION:
                probe = self._sendWithTimeout(
                    "Runtime.evaluate",
                    {
                        "expression": self._readyExpression(),
                        "returnByValue": True,
                    },
                    session_id,
                )
                probe_result = cast(Mapping[str, Any], probe.get("result")) if isinstance(probe.get("result"), Mapping) else {}
                ready = bool(probe_result.get("value"))
            if not ready:
                self._sendWithTimeout("Target.detachFromTarget", {"sessionId": session_id})
                return None
            match = EXT_ID_FROM_URL_RE.match(target["url"])
            extension_id = value.get("extension_id") if isinstance(value.get("extension_id"), str) else None
            result: ExtensionInjectionResult = {
                "source": "borrow",
                "extension_id": extension_id or (match.group(1) if match else None),
                "target_id": target["targetId"],
                "url": target["url"],
                "session_id": session_id,
            }
            return result, bool(value.get("has_tabs")), bool(value.get("has_debugger"))
        except BaseException:
            self._sendWithTimeout("Target.detachFromTarget", {"sessionId": session_id})
            raise
